File: packages/python-sdk/src/chromiumfish/launcher.py
"""Shared launch-argument construction for the sync/async wrappers."""
from __future__ import annotations

import os
from pathlib import Path
from typing import Any

# Flags that keep the GPU-less / SwiftShader path working and the persona
# engine happy. Mirrors the production launch_lean.sh defaults (minus anything
# that is now baked into the build / bundled addon).
BASE_ARGS: list[str] = [
    "--no-sandbox",
    "--no-zygote",
    "--disable-dev-shm-usage",
    "--use-gl=angle",
    "--use-angle=swiftshader",
    "--enable-unsafe-swiftshader",
]


def build_args(
    *,
    persona_seed: int | None = None,
    window_size: tuple[int, int] | None = (1920, 1080),
    extra_args: list[str] | None = None,
) -> list[str]:
    args = list(BASE_ARGS)
    if persona_seed is not None:
        args.append(f"--persona-seed={persona_seed}")
    if window_size is not None:
        args.append(f"--window-size={window_size[0]},{window_size[1]}")
    if extra_args:
        args.extend(extra_args)
    return args


def proxy_to_url(proxy: dict[str, Any] | None) -> str | None:
    """Flatten a Playwright proxy dict into a ``scheme://user:pass@host:port``
    URL for the egress probe. Returns None when no usable server is set."""
    if not proxy or not proxy.get("server"):
        return None
    server = proxy["server"]
    user, pwd = proxy.get("username"), proxy.get("password")
    if user and "://" in server:
        scheme, host = server.split("://", 1)
        return f"{scheme}://{user}:{pwd or ''}@{host}"
    return server


def resolve_timezone(
    timezone: str | None,
    *,
    proxy: dict[str, Any] | None,
    download: bool,
) -> str | None:
    """Interpret the wrapper's ``timezone`` option into a concrete IANA zone.

    * None       -> no timezone handling (returns None)
    * "auto"     -> probe the egress IP (through ``proxy``) and resolve it
                    against the downloadable ip2tz DB
    * "<IANA>"   -> used verbatim (no probe, no DB)
    """
    if not timezone:
        return None
    if timezone != "auto":
        return timezone
    from .ip2tz import resolve_timezone as _resolve  # lazy: avoids DB import cost
    return _resolve(proxy=proxy_to_url(proxy), download=download)


def launch_options(
    *,
    executable_path: Path,
    headless: bool,
    persona_seed: int | None,
    proxy: dict[str, Any] | None,
    window_size: tuple[int, int] | None,
    args: list[str] | None,
    extra: dict[str, Any],
    tz: str | None = None,
) -> dict[str, Any]:
    opts: dict[str, Any] = {
        "executable_path": str(executable_path),
        "headless": headless,
        "args": build_args(
            persona_seed=persona_seed, window_size=window_size, extra_args=args
        ),
    }
    if proxy is not None:
        opts["proxy"] = proxy
    opts.update(extra)
    if tz:
        # Chromium's ICU adopts TZ at process init — same mechanism the
        # production launch_lean.sh uses as the timezone source of truth.
        # Playwright's `env` replaces (not merges) the child environment, so
        # start from the current one. Our resolved TZ wins over any inherited.
        env = dict(os.environ)
        env.update(opts.get("env") or {})
        env["TZ"] = tz
        opts["env"] = env
    return opts

File: packages/js-sdk/src/launcher.ts
import { chromium, type Browser, type LaunchOptions } from "playwright-core";
import { binaryPath } from "./fetch.js";
import { resolveTimezone } from "./ip2tz.js";

/**
 * Flags that keep the GPU-less / SwiftShader path working and the persona
 * engine happy. Mirrors the production launch_lean.sh defaults (minus anything
 * baked into the build / bundled addon).
 */
export const BASE_ARGS: string[] = [
  "--no-sandbox",
  "--no-zygote",
  "--disable-dev-shm-usage",
  "--use-gl=angle",
  "--use-angle=swiftshader",
  "--enable-unsafe-swiftshader",
];

export interface ChromiumFishOptions extends Omit<LaunchOptions, "executablePath"> {
  /** String id for a stable, internally-consistent fingerprint persona. */
  personaSeed?: string;
  /** Run headless (SwiftShader). Defaults to true. */
  headless?: boolean;
  /** Window dimensions; defaults to [1920, 1080]. Pass null to omit. */
  windowSize?: [number, number] | null;
  /** Override the browser build version. */
  version?: string;
  /** Auto-download the build if missing. Defaults to true. */
  download?: boolean;
  /**
   * Timezone handling. `"auto"` probes the egress IP and resolves it against
   * the downloadable ip2tz DB; an IANA string (e.g. `"Europe/Berlin"`) is used
   * verbatim; omit to leave the timezone untouched.
   */
  timezone?: string;
}

/** Flatten a Playwright proxy option into a probe URL, or undefined. */
function proxyToUrl(proxy: LaunchOptions["proxy"]): string | undefined {
  if (!proxy?.server) return undefined;
  const { server, username, password } = proxy;
  if (username && server.includes("://")) {
    const [scheme, host] = server.split("://", 2);
    return `${scheme}://${username}:${password ?? ""}@${host}`;
  }
  return server;
}

export function buildArgs(opts: ChromiumFishOptions): string[] {
  const args = [...BASE_ARGS];
  if (opts.personaSeed !== undefined) args.push(`--persona-seed=${opts.personaSeed}`);
  const ws = opts.windowSize === undefined ? [1920, 1080] : opts.windowSize;
  if (ws) args.push(`--window-size=${ws[0]},${ws[1]}`);
  if (opts.args) args.push(...opts.args);
  return args;
}

/**
 * Launch ChromiumFish and return a standard Playwright `Browser`.
 *
 *   import { ChromiumFish } from "chromiumfish";
 *   const browser = await ChromiumFish({ personaSeed: "alpha-7", headless: true });
 */
export async function ChromiumFish(opts: ChromiumFishOptions = {}): Promise<Browser> {
  const { personaSeed, headless = true, windowSize, version, download = true, timezone, args, ...launch } = opts;
  const executablePath = await binaryPath(version, download);

  // Resolve the timezone before launch: "auto" -> egress IP via the ip2tz DB,
  // an IANA string -> used as-is. Inject as the TZ env var so Chromium's ICU
  // adopts it at process init (the production timezone source of truth).
  let tz: string | null = null;
  if (timezone) {
    tz = timezone === "auto" ? await resolveTimezone({ proxy: proxyToUrl(launch.proxy), download }) : timezone;
  }
  let env = launch.env;
  if (tz) {
    env = {
      ...(process.env as Record<string, string>),
      ...((launch.env as Record<string, string> | undefined) ?? {}),
      TZ: tz,
    };
  }

  return chromium.launch({
    executablePath,
    headless,
    args: buildArgs({ personaSeed, windowSize, args }),
    ...launch,
    ...(env ? { env } : {}),
  });
}
